pkg/client/introduce/client.go

```diff
@@ -9,220 +7,131 @@ SPDX-License-Identifier: Apache-2.0
 func (c *Client) SendRequest(to *introduce.PleaseIntroduceTo, myDID, theirDID string) error {
-	return c.handleOutbound(&introduce.Request{
+	_, err := c.service.HandleOutbound(service.NewDIDCommMsgMap(&introduce.Request{
 		Type:              introduce.RequestMsgType,
-		ID:                c.newUUID(),
 		PleaseIntroduceTo: to,
-	}, InvitationEnvelope{
-		Recps: []*introduce.Recipient{{MyDID: myDID, TheirDID: theirDID}},
-	})
-}
-
-// HandleRequest is a helper function to prepare the right protocol dependency interface.
-// It can be executed after receiving a Request action message (the client does not have a public Invitation)
-func (c *Client) HandleRequest(msg service.DIDCommMsg, to *introduce.To, recipient *introduce.Recipient) error {
-	thID, err := msg.ThreadID()
-	if err != nil {
-		return fmt.Errorf("handle request threadID: %w", err)
-	}
-
-	return c.saveInvitationEnvelope(thID, InvitationEnvelope{
-		Recps: []*introduce.Recipient{{To: to}, recipient},
-	})
-}
-
-// HandleRequestWithInvitation is a helper function to prepare the right protocol dependency interface.
-// It can be executed after receiving a Request action message (the client has a public Invitation)
-func (c *Client) HandleRequestWithInvitation(msg service.DIDCommMsg,
-	inv *didexchange.Invitation, to *introduce.To) error {
-	thID, err := msg.ThreadID()
-	if err != nil {
-		return fmt.Errorf("handle request with invitation threadID: %w", err)
-	}
+	}), myDID, theirDID)
 
-	return c.saveInvitationEnvelope(thID, InvitationEnvelope{
-		Inv:   inv,
-		Recps: []*introduce.Recipient{{To: to}},
-	})
+	return err
 }
 
-func (c *Client) handleOutbound(msg interface{}, o InvitationEnvelope) error {
-	payload, err := json.Marshal(msg)
-	if err != nil {
-		return fmt.Errorf("marshal outbound msg: %w", err)
-	}
-
-	didMsg, err := service.ParseDIDCommMsgMap(payload)
-	if err != nil {
-		return fmt.Errorf("new outbound DIDCommMsg msg: %w", err)
-	}
-
-	thID, err := didMsg.ThreadID()
-	if err != nil {
-		return fmt.Errorf("outbound threadID: %w", err)
-	}
-
-	if _, ok := msg.(*introduce.Request); !ok {
-		if err := c.saveInvitationEnvelope(thID, o); err != nil {
-			return err
-		}
-	}
-
-	return c.service.HandleOutbound(didMsg, o.Recps[0].MyDID, o.Recps[0].TheirDID)
-}
-
-// InvitationEnvelope keeps the information needed for sending a proposal
-type InvitationEnvelope struct {
-	// Invitation must be set when we have a public Invitation
-	Inv *didexchange.Invitation `json:"invitation,omitempty"`
-	// Recipients contain one or two elements
-	// one element - for the public Invitation, otherwise two elements
-	Recps []*introduce.Recipient `json:"recipients,omitempty"`
+// ContinueWithInvitation is used when introducee wants to provide invitation.
+// NOTE: For async usage. Introducee can provide invitation only after receiving ProposalMsgType
+func (c *Client) ContinueWithInvitation(piID string, inv *didexchange.Invitation) error {
```

Comment:
What's the client API which provides piID in async usecase ?

Reply:
See `UnfinishedActions` function